### PR TITLE
Android embedding API updates for plugin ecosystem

### PR DESCRIPTION
**Pull request by matthew-carroll**
- [x] `FlutterPluginBinding` is a facade, `getFlutterEngine()` is deprecated

- [x] `Lifecycle` is split between `ActivityAware` and `ServiceAware`, `FlutterEngine#getLifecycle()` is deprecated

- [x] `Activity` instance state saving callbacks are forwarded to plugins

### CODE REVIEW

**Comment by matthew-carroll** [comment]
@amirh @mklim @cyanglaz @bparrishMines @collinjackson @kroikie Can you all please take a look at `FlutterPluginBinding` and let me know if you have any thoughts about additions/removals of accessors as a facade for plugins? I went with the few things that seemed obvious and necessary. We should try to reach consensus on that API surface ASAP. Thanks.

**Comment by mklim** [comment]
The facade API LG to me, I didn't notice anything missing. I could very likely be forgetting an edge case though.

**Comment by tvolkert** [comment]
API LGTM.  We can always add more to the surface if needed later.

**Comment by matthew-carroll** [comment]
I added JVM tests to ensure that the new plugin facade references are not null when the binding is given to plugins. I also verified that state saving callbacks are forwarded to plugins that register for them.

**Comment by matthew-carroll** [comment]
@mklim would you mind taking a look at this PR? I'd like to merge this in before beginning work on creating a custom Lifecycle interface.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterAndroidComponentTest.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    when(mockFlutterJni.isAttached()).thenReturn(true);
+    FlutterEngine cachedEngine = spy(new FlutterEngine(RuntimeEnvironment.application, mockFlutterLoader, mockFlutterJni));
+    FlutterEngineCache.getInstance().put("my_flutter_engine", cachedEngine);
+
+    // Add mock plugin.
+    FlutterPlugin mockPlugin = mock(FlutterPlugin.class);
+    cachedEngine.getPlugins().add(mockPlugin);
+
+    // Create a fake Host, which is required by the delegate.
+    FlutterActivityAndFragmentDelegate.Host mockHost = mock(FlutterActivityAndFragmentDelegate.Host.class);
```
Comment (mklim) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterAndroidComponentTest.java]: readability nit: It would probably be better to create a stub than a mock here. Subjective but stubs are generally easier to reason about than a series of `when...thenReturn()` statements.

**Comment by kroikie** [comment]
LGTM from FlutterFire end.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java [hunk 2]
```diff
@@ -223,7 +214,7 @@ private void setupFlutterEngine() {
     // FlutterView.
     Log.d(TAG, "No preferred FlutterEngine was provided. Creating a new FlutterEngine for"
         + " this FlutterFragment.");
-    flutterEngine = new FlutterEngine(host.getContext());
+    flutterEngine = new FlutterEngine(host.getContext(), host.getFlutterShellArgs().toArray());
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java]: Yep

File: shell/platform/android/test/io/flutter/embedding/android/FlutterAndroidComponentTest.java (hunk 1 above)
Comment (matthew-carroll) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterAndroidComponentTest.java]: Sure. I switched out those mocks for fake implementations.

File: shell/platform/android/io/flutter/embedding/engine/plugins/activity/ActivityPluginBinding.java [hunk 3]
```diff
@@ -23,6 +25,12 @@
   @NonNull
   Activity getActivity();
 
+  /**
+   * Returns the {@link Lifecycle} associated with the attached {@code Activity}.
+   */
+  @NonNull
+  Lifecycle getLifecycle();
```
Comment (amirh) [inline shell/platform/android/io/flutter/embedding/engine/plugins/activity/ActivityPluginBinding.java]: Lifecycle should be fetched through the flutter_android_lifecycle plugin, and not available directly here
